### ISSUE TITLE
experts based on Phi 3 model with PEFT library

### ISSUE THREAD

**Issue by AliEdalat** [microsoft/mttl#124]
Hello,

Thank you for your interesting research work.

I have 10 experts trained based on the Phi 3 model (datasets selected based on paper cluttering). I have used the TRL and PEFT libraries for training, ensuring the checkpoint structures are suitable for these libraries.

In training the experts, I used LoRA in 4-bit quantized mode. Additionally, I utilized the o and kqv attention in each layer during training.

I would like to know how I can use your code to execute Arrow for merging these experts for each token in every model layer.

I have some errors in the code.

please explain step by step. I am a beginner in this field.

Thank you, and I would appreciate your response.

**Comment by sordonia** [comment]
@pclucas14 can you help with this issue?

I also have a PR that creates a library compatible with our library out of PEFT adapters, maybe we can merge that and write an example on how to run Arrow using experts trained with TRL and PEFT

Thank you

**Comment by sordonia** [comment]
@AliEdalat would you be able to share with us the name of your PEFT adapters?

**Comment by TheTahaaa** [comment]
Hey @sordonia,

I'm @AliEdalat's teammate, and have implemented the Arrow algorithm for our project using the PEFT library, achieving promising results. To do this, I modified the forward path in the [bnb.py file](https://github.com/huggingface/peft/blob/main/src/peft/tuners/lora/bnb.py) (since we use QLoRA in our workflow) and made adjustments to methods in the [peft_model.py file](https://github.com/huggingface/peft/blob/main/src/peft/peft_model.py), along with some other PEFT library files.

I thought you might be interested in a dedicated PEFT adapter for the Arrow algorithm, similar to existing ones like[Poly and MHR](https://huggingface.co/docs/peft/en/package_reference/poly). Let me know if you'd like any help on this!

Thanks!

**Comment by sordonia** [comment]
It'd be great! Do you plan to open a PR into PEFT? I'd be willing to help

**Comment by TheTahaaa** [comment]
> It'd be great! Do you plan to open a PR into PEFT? I'd be willing to help

Yes, I'd be glad to contribute!
@sordonia 

Which branch should I open a PR on?

**Comment by sordonia** [comment]
Just to clarify my understanding: you are planning to PR your work into the huggingface PEFT library, right? Or into MTTL library?

**Comment by TheTahaaa** [comment]
> Just to clarify my understanding: you are planning to PR your work into the huggingface PEFT library, right? Or into MTTL library?

Yes! HuggingFace PEFT library.

I initially asked about the branch because I thought there might already be a feature request for implementing Arrow in PEFT. Should I go ahead and open a feature request for this, or would you prefer to handle it?

(Sorry for any ambiguity!)

**Comment by TheTahaaa** [comment]
> Just to clarify my understanding: you are planning to PR your work into the huggingface PEFT library, right? Or into MTTL library?

Yes! HuggingFace PEFT library.

I initially asked about the branch because I thought there might already be a feature request for implementing Arrow in PEFT. Should I go ahead and open a feature request for this, or would you prefer to handle it?

(Sorry for any ambiguity!)

**Comment by pclucas14** [comment]
Hi, 

Super exciting to hear you are getting nice results! Please go ahead and PR into PEFT. I am happy to help if needed, but you probably have a better intuition on how PEFT works than us :)

**Comment by sordonia** [comment]
Yes please go ahead an open a feature request, we can jump in the PR when needed!

**Comment by sordonia** [comment]
(we merged our own PEFT support in `examples/create_arrow_model.py` curious if it works with your PEFT experts :)

**Comment by TheTahaaa** [comment]
@sordonia @pclucas14 I'll do it for sure. However, I need to first make sure my code can almost reproduce the result on the zero-shot datasets, like Piqa, as you wrote on the paper.

Furthermore, I think I need to refactor the code, so the Arrow can be added as a Tuner Class in the PEFT library, similar to Polytropon and π-Tuning.

**Comment by sordonia** [comment]
@TheTahaaa how is it going? :)

**Comment by TheTahaaa** [comment]
Hey @sordonia!

I’ve tested my implementation of the Arrow algorithm on the benchmarks from your paper (PIQA, BOOLQ, etc.) using the Phi3 model, and the results were quite promising, confirming that the algorithm works well.

I was just about to start working on the PR—so thanks for the reminder! 😁

Apologies for the delay; we only have access to a single RTX 3090 (24GB), which made testing the algorithm on the benchmark datasets a bit challenging.

**Comment by sordonia** [comment]
Awesome!! I am really looking forward to know more about the setting and no problem at all, I was just curious so I pinged, let me know the progress and how we can help

**Comment by TheTahaaa** [comment]
I'll definitely keep you updated on the progress and share more details about the setting soon.

To make things more transparent and easier to follow, I was thinking we could set up this on Git. That way, you can easily track changes and progress as we go. 

Let me know if that works for you, and I’d be happy to set it up!

**Comment by sordonia** [comment]
sure! thanks

**Comment by TheTahaaa** [comment]
Hi @sordonia,

I’ve forked the PEFT repository and added you to the repo. Before we dive in, I think it would be helpful to align on the approach we’ll take to implement this, as there seem to be at least two possible ways to proceed.

What’s the best way to get in touch to discuss this—email or another platform? Let me know what works best for you!

**Comment by TheTahaaa** [comment]
Hey there 👋🏻, 

I just submitted a [PR](https://github.com/huggingface/peft/pull/2644) to the PEFT library, adding support for Arrow as well as our method **GenKnowSub**, introduced in our [paper](https://arxiv.org/abs/2505.10939), which has been accepted to ACL 2025 Main Conference. GenKnowSub builds on Arrow by first subtracting a general-domain LoRA from each task-specific adapter to reduce redundancy and enhance specialization before routing. This refinement consistently improves performance in our experiments.

I’ve made sure everything aligns with PEFT’s standards, and I’m hopeful the PR will be reviewed and integrated soon.

Would love to hear your thoughts on our work or the integration if you get a chance to take a look at either the paper or the PR

Thanks in advance! 🙌🏻

**Comment by pclucas14** [comment]
Congrats on the work! I really like the idea of substracting a common/general adapter to get better prototypes :)
